src/ot_orchestration/operators/gce.py

```diff
@@ -0,0 +1,733 @@
+
+    If the operation is successful, it will return its result.
+    If the operation ends with an error, an exception will be raised.
+    If there were any warnings during the execution of the operation
+    they will be logged.
+
+    Args:
+        operation: a long-running operation you want to wait on.
+        verbose_name: (optional) a more verbose name of the operation,
+            used only during error and warning reporting.
+        timeout: how long (in seconds) to wait for operation to finish.
+            If None, wait indefinitely.
+        log: (optional) a logger to use for logging.
+
+    Returns:
+        Whatever the operation.result() returns.
+
+    Raises:
+        This method will raise the exception received from `operation.exception()`
+        or RuntimeError if there is no exception set, but there is an `error_code`
+        set for the `operation`.
+
+        In case of an operation taking longer than `timeout` seconds to complete,
+        a `concurrent.futures.TimeoutError` will be raised.
+    """
+    result = operation.result(timeout=timeout)
+
+    if operation.error_code:
+        log.error(
+            f"Error during {verbose_name}: [Code: {operation.error_code}]: {operation.error_message}",
+        )
+        raise operation.exception() or RuntimeError(operation.error_message)
+
+    if operation.warnings:
+        log.warning(f"Warnings during {verbose_name}")
+        for warning in operation.warnings:
+            log.warning(f"{warning.code}: {warning.message}")
+
+    return result
+
+
+class RateLimitedLoggingClient(logging_v2.Client):
+    """Client for the Google Cloud Logging service with rate limiting.
+
+    This client will wait for a minimum interval between requests to avoid
+    hitting the rate limits of the Google Cloud Logging service.
+
+    We are hitting logging API rate limits when we are trying to list log entries
+    to copy them to the airflow logs, as there are numerous concurrent requests
+    when running all of PIS steps in parallel.
+
+    This may delay the logs from being copied to the Airflow logs for steps with
+    a large number of log entries, but it will prevent the rate limit errors.
+    """
+
+    def __init__(self, log: logging.Logger, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.log = log
+        self.request_interval = LOGGING_REQUEST_INTERVAL
+
+    def list_entries(self, *args, **kwargs):
+        """List log entries and retries request that get rate-limited."""
+        entries = None
+
+        while True:
+            try:
+                entries = super().list_entries(*args, **kwargs)
+                break
+            except ResourceExhausted:
+                self.log.warning(
+                    "Rate limit for logging api exceeded, waiting for %d seconds",
+                    self.request_interval,
+                )
+                time.sleep(self.request_interval)
+                self.request_interval *= 2
```

Comment:
I did this on purpose, because if the request fails for a quota limit, it might be necessary to wait up to a minute. We perform less requests if exponentially increasing than if we check every x seconds.

Reply:
I agree, but then again you do not recover from it after successful attempt, which will prolong the next consecutive checks.

Reply:
Oh, that is a very good point! I'll reset the interval when a request is successful.